girder_worker/docker/tasks/__init__.py

```diff
@@ -49,13 +49,18 @@ def _run_container(image, container_args,  **kwargs):
     client = docker.from_env(version='auto')
 
     runtime = kwargs.pop('runtime', None)
+    origRuntime = runtime
     if runtime is None and nvidia.is_nvidia_image(client.api, image):
         runtime = 'nvidia'
 
     logger.info('Running container: image: %s args: %s runtime: %s kwargs: %s'
                 % (image, container_args, runtime, kwargs))
     try:
-        return client.containers.run(image, container_args, runtime=runtime, **kwargs)
+        try:
+            return client.containers.run(image, container_args, runtime=runtime, **kwargs)
+        except APIError:
```

Comment:
I'm assuming this is not the exception type raised if, e.g. the docker container crashes or returns non-zero, right?

Reply:
It looks like `APIError` gets subclassed as the `NotFound` error as well, which can get raised in other instances.  The more specific error is `APIError: 400 Client Error: Bad Request ("Unknown runtime specified nvidia")`.  In the event that a different error is received, this will try again (without nvidia), which will likely result in the same failure a second time.

Reply:
To follow up on this, the things I asked about raise the more generic `DockerException` type, of which `APIError` is a subclass.